### ISSUE TITLE
Overlay has maps that grow without bound

### ISSUE THREAD

**Issue by jeffallen** [dedis/onet#363]
This is an issue to gather notes related to memory leak investigations, and possible code to detect and avoid them.

Observations:
1. there is one Overlay structure per server, and servers are long-lived, and must not leak resources
2. Overlay has several maps in it
3. Thus: the balance of adds/deletes to these maps must be maintained, and our testing infrastructure must enforce this balance.

Todo:
1. modify local.CloseAll to verify that all the maps in all the Overlays are empty before closing. There are lots: trees, entityLists, instances, protocolInstances, instancesInfo, pendingTreeMarshal, pendingConfigs. Also: slice pendingMsg, which must be empty.
1. set up a simulation which repeats operations many times, and measure the max RAM size for the simulation process. It will likely be monotonically increasing. Fix the leaks until it's no longer doing so.

Problems:
1. what is the purpose of Overlay.instanceInfo? When is it no longer needed? Answer: If a message arrives for a Protocol that has already called Done(), you do not want to create a new ProtocolInstance, you need to know to drop the message. Possible solution: an expiration time on entries in instanceInfo, with an explicit call to expire them in CloseAll?

**Comment by kc1212** [comment]
![image](https://user-images.githubusercontent.com/1093806/37202734-a10977f4-238b-11e8-87cd-634b55f0a9e9.png)

Did a little analysis with ftcosi. Basically ran it for 1000 rounds and did this
```
func PrintMemUsage() {
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	fmt.Printf("Alloc,HeapSys:%d,%d\n", m.Alloc, m.HeapSys)
}
```
at the end of every round.

Clearly the memory is growing. Some of it might be related to the monitor, but with the fix, we should see something much better.

**Comment by jeffallen** [comment]
CC @ineiti @nikkolasg

**Comment by kc1212** [comment]
Not too big of a difference unfortunately. Perhaps we can run a profiler to figure out the hot spots? I'll try to see how that can be done.
![image](https://user-images.githubusercontent.com/1093806/37203746-ff115f94-238e-11e8-83c2-f7d8908c299d.png)

**Comment by jeffallen** [comment]
We are clearly on the right track though, because the growth rate changed. So now the heavy lifting of going through the rest of the maps in Overlay and getting their lifecycles under control.

**Comment by jeffallen** [comment]
I am now investigating Overlay.trees. It seems likely we'll need to add something to the tree cache pointing to which connection we discovered the tree on and free it when the connection goes down.

**Comment by ineiti** [comment]
I think that the `Overlay.trees` is not the culprit here, because we always use the same tree, and then it _should_ only use one copy. Also, when a connection goes down, we should keep the tree, in case that it is requested again. Else we'll have to send/receive the tree again every time we make a new connection!

**Comment by kc1212** [comment]
did some profiling using pprof - [profile001.pdf](https://github.com/dedis/onet/files/1797392/profile001.pdf)

it shows that `RegisterChannel` is using over half of the memory, maybe we need to make sure these channels are deleted (I haven't check). but I'm also not sure how useful this graph is. I think it shows the amount of allocation rather than like a dump of the heap at the very end.

**Comment by kc1212** [comment]
The channels are a part of TNI, so the GC should sweep them

**Comment by kc1212** [comment]
It's possible to get a heapdump actually, using https://golang.org/pkg/runtime/debug/#WriteHeapDump

I just did it after leaving the simulation to run for 100 rounds and it creates a 400 MB file. It (almost) literally dumps the heap, so it's not easy to analyse what's actually in it, I couldn't find any tools. It's in a format specified by go developers https://github.com/golang/go/wiki/heapdump15-through-heapdump17

**Comment by kc1212** [comment]
The profile visualisation https://github.com/dedis/onet/files/1797392/profile001.pdf is, in fact, valid. A small test shows that the memory profile is quite different in the two examples below:
```
package main

import (
	"runtime/debug"

	"github.com/pkg/profile"
)

func main() {
	defer profile.Start(profile.MemProfile).Stop()
	var a []int
	var i int
	for {
		if i > 1024*1024 {
			break
		}
		a = append(a, i)
		i++
	}
	debug.FreeOSMemory() // the other version doesn't have this line
}
```

[profile002_nogc.pdf](https://github.com/dedis/onet/files/1801964/profile002_nogc.pdf)
[profile002_gc.pdf](https://github.com/dedis/onet/files/1801965/profile002_gc.pdf)

**Comment by kc1212** [comment]
Fixed by #364
